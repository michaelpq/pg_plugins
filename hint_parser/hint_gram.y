%{
/*-------------------------------------------------------------------------
 *
 * hint_gram.y				- Parser for query hints
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  src/backend/replication/hint_gram.y
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "nodes/pg_list.h"
#include "hint_parser.h"

/* Result of parsing is returned in one of these two variables */
List	   *hint_parse_result;	/* list of hints */
char	   *hint_parse_error_msg;

static HintConfigData *create_hint_item(const char *hint_name,
										const char *hint_contents);

/*
 * Bison doesn't allocate anything that needs to live across parser calls,
 * so we can easily have it use palloc instead of malloc.  This prevents
 * memory leaks if we error out during parsing.
 */
#define YYMALLOC palloc
#define YYFREE   pfree

%}

%expect 0
%name-prefix="hint_yy"

%union
{
	char	   *str;
	List	   *list;		/* should be a list of HintConfigData */
	HintConfigData *item;
}

%token <str> NAME

%type <list> hint_list

%type <str> hint_name hint_content
%type <item> hint_item
%type <list> result

%start result

%%
result:
		hint_list				{ hint_parse_result = $1; }
	;

hint_list:
		hint_item					{ $$ = list_make1($1); }
		| hint_list hint_item		{ $$ = lappend($1, $2); }
	;

hint_item:
		hint_name '(' hint_content ')'
		{
			$$ = create_hint_item($1, $3);
		}
	;

hint_name:			NAME { $$ = $1; };
hint_content:		NAME { $$ = $1; };

%%

static HintConfigData *
create_hint_item(const char *hint_name, const char *hint_contents)
{
	HintConfigData *res = palloc0(sizeof(HintConfigData));

	res->name = pstrdup(hint_name);
	res->contents = pstrdup(hint_contents);

	return res;
}
