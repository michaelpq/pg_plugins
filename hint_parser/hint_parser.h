/*-------------------------------------------------------------------------
 *
 * hint_parser.h
 *
 * Copyright (c) 1996-2025, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  hint_parser/hint_parser.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef HINT_PARSER_H
#define HINT_PARSER_H

#include "nodes/pg_list.h"

/* silence -Wmissing-variable-declarations */
extern int	hint_yychar;
extern int	hint_yynerrs;

typedef struct HintConfigData
{
	char	   *name;
	char	   *contents;
} HintConfigData;

/* internal scan state of lexer */
typedef struct HintScanStateData
{
	int			depth;
} HintScanStateData;

/* Abstract type for lexer's internal state */
typedef struct HintScanStateData *HintScanState;

/* communication variables for parsing */
extern PGDLLIMPORT List *hint_parse_result;
extern PGDLLIMPORT char *hint_parse_error_msg;

/*
 * Internal functions for parsing query strings in hint_gram.y and
 * hint_scanner.l.
 */
extern int	hint_yyparse(void);
extern int	hint_yylex(void);
extern void hint_yyerror(const char *str);
extern void hint_scanner_init(const char *str);
extern void hint_scanner_finish(void);

#endif							/* HINT_PARSER_H */
