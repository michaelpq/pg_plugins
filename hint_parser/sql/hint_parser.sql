CREATE EXTENSION hint_parser;
-- Hint in comments
select hint_parser('/*+h1(h2)*/');
select hint_parser('/*+h1(h2) */');
select hint_parser('/*+h1(h2) */');
select hint_parser('/*+ h1(h2) h3(h4) */');
select hint_parser('/*+ h1(h2) h3(h4) */ /*+ h5(h2) h3(h4) */ /*+ h1(h2) h3(h4) */ ');
