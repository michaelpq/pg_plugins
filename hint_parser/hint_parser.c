/*-------------------------------------------------------------------------
 *
 * hint_parser.c
 *		Set of functions to parse query hints from query strings.
 *
 * Copyright (c) 1996-2025, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  hint_parser/hint_parser.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "hint_parser.h"

#include "fmgr.h"
#include "funcapi.h"
#include "utils/builtins.h"

PG_MODULE_MAGIC;

/*
 * This is the hint_parser function.
 */
PG_FUNCTION_INFO_V1(hint_parser);

/*
 * hint_parser
 *
 * Parse and return as a SRF all the hints found in a given string.  The
 * content given in input should be extracted from a query.
 */
Datum
hint_parser(PG_FUNCTION_ARGS)
{
	char	   *string = text_to_cstring(PG_GETARG_TEXT_PP(0));
	int			parse_rc;
	ListCell   *item;
	List	   *hints = NIL;
	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;

	InitMaterializedSRF(fcinfo, 0);

	/* Results from the parsing */
	hint_parse_result = NULL;
	hint_parse_error_msg = NULL;

	/* Parse the query */
	hint_scanner_init(string);
	parse_rc = hint_yyparse();
	hint_scanner_finish();

	if (parse_rc != 0 || hint_parse_result == NULL)
	{
		if (hint_parse_error_msg)
			elog(ERROR, "%s", hint_parse_error_msg);
		else
			elog(ERROR, "hint parser failed");
	}

	hints = hint_parse_result;

	/* Grab and store all the hint items */
	foreach(item, hints)
	{
		HintConfigData *hint = lfirst(item);
		Datum		values[2];
		bool		nulls[2];

		values[0] = CStringGetTextDatum(hint->name);
		values[1] = CStringGetTextDatum(hint->contents);
		memset(nulls, 0, sizeof(nulls));

		tuplestore_putvalues(rsinfo->setResult, rsinfo->setDesc, values, nulls);
	}

	return (Datum) 0;
}
