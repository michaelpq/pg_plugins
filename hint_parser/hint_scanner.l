%top{
/*-------------------------------------------------------------------------
 *
 * hint_scanner.l
 *	  a lexical scanner for query hints
 *
 * Portions Copyright (c) 1996-2025, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  hint_parser/hint_scanner.l
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "lib/stringinfo.h"
#include "nodes/pg_list.h"

/*
 * NB: include hint_gram.h only AFTER including hint.h, because hint.h
 * includes node definitions needed for YYSTYPE.
 */
#include "hint_parser.h"
#include "hint_gram.h"
}

%{
/* Avoid exit() on fatal scanner errors (a bit ugly -- see yy_fatal_error) */
#undef fprintf
#define fprintf(file, fmt, msg)  fprintf_to_ereport(fmt, msg)

static void
fprintf_to_ereport(const char *fmt, const char *msg)
{
	ereport(ERROR, (errmsg_internal("%s", msg)));
}

/* Handles to the buffer that the lexer uses internally */
static YY_BUFFER_STATE scanbufhandle;

/* LCOV_EXCL_START */

%}

%option 8bit
%option never-interactive
%option nodefault
%option noinput
%option nounput
%option noyywrap
%option warn
%option prefix="hint_yy"

space			[ \t\n\r\f\v]

%%

{space}+	{ /* ignore */ }

[A-Za-z0-9]+	{
			hint_yylval.str = pstrdup(yytext);
			return NAME;
		}

"("		{ return '('; }
")"		{ return ')'; }

<<EOF>>		{
			yyterminate();
		}

.		{ /* ignore */ }

%%

/* LCOV_EXCL_STOP */

/* Needs to be here for access to yytext */
void
hint_yyerror(const char *message)
{
	/* report only the first error in a parse operation */
	if (hint_parse_error_msg)
		return;
	if (yytext[0])
		hint_parse_error_msg = psprintf("%s at or near \"%s\"",
						message, yytext);
	else
		hint_parse_error_msg = psprintf("%s at end of input",
						message);
}

void
hint_scanner_init(const char *str)
{
	Size		slen = strlen(str);
	char	   *scanbuf;

	/*
	 * Might be left over after ereport()
	 */
	if (YY_CURRENT_BUFFER)
		yy_delete_buffer(YY_CURRENT_BUFFER);

	/*
	 * Make a scan buffer with special termination needed by flex.
	 */
	scanbuf = (char *) palloc(slen + 2);
	memcpy(scanbuf, str, slen);
	scanbuf[slen] = scanbuf[slen + 1] = YY_END_OF_BUFFER_CHAR;
	scanbufhandle = yy_scan_buffer(scanbuf, slen + 2);

	/* Make sure we start in proper state */
	BEGIN(INITIAL);
}

void
hint_scanner_finish(void)
{
	yy_delete_buffer(scanbufhandle);
	scanbufhandle = NULL;
}
