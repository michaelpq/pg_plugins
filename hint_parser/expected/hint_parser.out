CREATE EXTENSION hint_parser;
-- Hint in comments
select hint_parser('/*+h1(h2)*/');
WARNING:  Hint found: name h1 contents h2
 hint_parser 
-------------
 
(1 row)

select hint_parser('/*+h1(h2) */');
WARNING:  Hint found: name h1 contents h2
 hint_parser 
-------------
 
(1 row)

select hint_parser('/*+h1(h2) */');
WARNING:  Hint found: name h1 contents h2
 hint_parser 
-------------
 
(1 row)

select hint_parser('/*+ h1(h2) h3(h4) */');
WARNING:  Hint found: name h1 contents h2
WARNING:  Hint found: name h3 contents h4
 hint_parser 
-------------
 
(1 row)

select hint_parser('/*+ h1(h2) h3(h4) */ /*+ h5(h2) h3(h4) */ /*+ h1(h2) h3(h4) */ ');
WARNING:  Hint found: name h1 contents h2
WARNING:  Hint found: name h3 contents h4
WARNING:  Hint found: name h5 contents h2
WARNING:  Hint found: name h3 contents h4
WARNING:  Hint found: name h1 contents h2
WARNING:  Hint found: name h3 contents h4
 hint_parser 
-------------
 
(1 row)

